rust/lance/src/io/exec/pushdown_scan.rs

```diff
@@ -0,0 +1,1394 @@
+}
+
+impl DisplayAs for LancePushdownScanExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                let columns = self
+                    .projection
+                    .fields
+                    .iter()
+                    .map(|f| f.name.as_str())
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                write!(
+                    f,
+                    "LancePushdownScan: uri={}, projection=[{}], predicate={}, row_id=[{}], ordered={}",
+                    self.dataset.data_dir(),
+                    columns,
+                    self.predicate,
+                    self.config.with_row_id,
+                    self.config.ordered_output
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct FragmentScanner {
+    fragment: FileFragment,
+    projection: Arc<Schema>,
+    predicate_projection: Arc<Schema>,
+    predicate: Expr,
+    config: ScanConfig,
+    reader: FragmentReader,
+    stats: Option<RecordBatch>,
+}
+
+impl FragmentScanner {
+    pub async fn open(
+        fragment: Fragment,
+        dataset: Arc<Dataset>,
+        projection: Arc<Schema>,
+        predicate_projection: Arc<Schema>,
+        predicate: Expr,
+        config: ScanConfig,
+    ) -> Result<Self> {
+        let fragment = FileFragment::new(dataset.clone(), fragment);
+
+        // We will call the reader with projections. In order for this to work
+        // we must ensure that we open the fragment with the maximal schema.
+        let mut reader = fragment.open(dataset.schema()).await?;
+        if config.make_deletions_null {
+            reader.with_make_deletions_null();
+        }
+
+        // We only need the statistics for the predicate projection.
+        let stats = reader.read_page_stats(Some(&predicate_projection)).await?;
+
+        Ok(Self {
+            fragment,
+            projection,
+            predicate_projection,
+            predicate,
+            config,
+            reader,
+            stats,
+        })
+    }
+
+    pub async fn scan(self) -> Result<impl Stream<Item = Result<RecordBatch>> + 'static + Send> {
+        let batch_readahead = self.config.batch_readahead;
+        let simplified_predicates = self.simplified_predicates()?;
+        let ordered_output = self.config.ordered_output;
+
+        let scanner = Arc::new(self);
+
+        let stream = futures::stream::iter(simplified_predicates.into_iter().enumerate()).map(
+            move |(batch_id, predicate)| {
+                let scanner_ref = scanner.clone();
+                async move { scanner_ref.read_batch(batch_id, predicate).await }
+            },
+        );
+
+        let stream = if ordered_output {
+            stream
+                .buffered(batch_readahead)
+                .try_filter_map(|res| futures::future::ready(Ok(res)))
```

Comment:
I'm a little confused what this is achieving

Reply:
Yeah it's not obvious. `res` is `Option<RecordBatch>` where it is `None` if we didn't read any data for a batch (stats told us we could skip), so the `try_filter_map()` is removing the `None` and unwrapping the `Some`s. I've added a comment and a type signature for `res`.